File: web/src/status.ts
// The console's shared vocabulary: workflow statuses, the phase path the thread
// is drawn along, attempt budgets, and the two derivations the console makes
// from the fields the API exposes today.
//
// Statuses and their pill variants come from docs/design/web-console/
// specification.md §3.3; the phase path and its labels from §2.6.
import type { Workflow, WorkflowEvent } from "./api";
import { stripAnsi } from "./ansi";

export const PHASES = [
  "prepare", "plan", "implement", "pipeline", "review",
  "push", "pr", "checks", "human", "merge", "done",
] as const;

export type Phase = (typeof PHASES)[number];

export const PHASE_LABEL: Record<Phase, string> = {
  prepare: "Prepare",
  plan: "Plan",
  implement: "Implement",
  pipeline: "Pipeline",
  review: "AI review",
  push: "Push",
  pr: "Open PR",
  checks: "GitHub checks",
  human: "Human gate",
  merge: "Merge",
  done: "Complete",
};

export type PillVariant = "run" | "ok" | "bad" | "warn" | "wait" | "idle";

export type StatusMeta = {
  label: string;
  variant: PillVariant;
  /** Whether the pill's dot pulses — reserved for work actively in flight. */
  pulse: boolean;
  /** Where this status sits on the phase path, or null for terminal-bad states. */
  phase: Phase | null;
};

// The Go V1 orchestrator writes exactly thirteen of these
// (orchestrator/internal/server/status.go's Status type is the source of
// truth): offered, preparing, planning, waiting_github_checks, delivering,
// waiting_human, waiting_ai_review, repairing, pipeline_failed and the four
// terminal statuses. `planning` (#351) is a real, wired status: a project that
// opts into requirePlanning sits here while its planner execution runs,
// before ever reaching `preparing`. `delivering` (#359), `waiting_ai_review`
// (#356), `repairing` (#357), `pipeline_failed` (#359) and `waiting_human`
// (the human-approval gate, resolved by SubmitHumanDecision in
// orchestrator/internal/server/management.go) are likewise real, wired
// statuses -- see status.go:16-137 for each one's full semantics; the
// summary that matters here is that all five hold the project lock and are
// active work in progress, exactly like `preparing`/`planning`, so none of
// them belongs in TERMINAL_STATUSES below on that basis alone.
// `pipeline_failed` is the one deliberate exception: status.go documents it
// as "a brief hand-off to the repair-or-block decision", and the console
// treats that hand-off as effectively terminal for the *display* purposes
// TERMINAL_STATUSES serves (the active count, the workflow list's
// active/terminal filter) even though the Go orchestrator's own
// moirai_active_workflows gauge counts it as active until the hand-off
// resolves -- see TERMINAL_STATUSES's comment below. Every other entry this
// table used to carry -- implementing, local_pipeline, ai_review, pushing,
// pr_created, merging, recovering -- named phases of a pipeline the deleted
// internal/workflow package (#247) modelled but the Go orchestrator never
// implemented, and were pruned in #265: statusMeta's fallback below covers a
// status this table does not recognise, which is what every one of those
// would have hit in production anyway.
export const STATUS_META: Record<string, StatusMeta> = {
  offered: { label: "Offered", variant: "idle", pulse: false, phase: "prepare" },
  planning: { label: "Planning", variant: "run", pulse: true, phase: "plan" },
  preparing: { label: "Preparing", variant: "run", pulse: true, phase: "prepare" },
  waiting_github_checks: { label: "Waiting on checks", variant: "wait", pulse: false, phase: "checks" },
  delivering: { label: "Delivering", variant: "run", pulse: true, phase: "pr" },
  waiting_human: { label: "Needs decision", variant: "wait", pulse: false, phase: "human" },
  waiting_ai_review: { label: "In AI review", variant: "run", pulse: true, phase: "review" },
  repairing: { label: "Repairing", variant: "warn", pulse: true, phase: "implement" },
  pipeline_failed: { label: "Pipeline failed", variant: "warn", pulse: false, phase: "pipeline" },
  completed: { label: "Delivered", variant: "ok", pulse: false, phase: "done" },
  blocked: { label: "Blocked", variant: "bad", pulse: false, phase: null },
  failed: { label: "Failed", variant: "bad", pulse: false, phase: null },
  cancelled: { label: "Cancelled", variant: "idle", pulse: false, phase: null },
};

// `pipeline_failed` is included here even though status.go's own
// terminalStatuses (the list its Prometheus active-workflow gauge and
// project-lock bookkeeping use) deliberately excludes it -- a run sitting at
// `pipeline_failed` still holds its project lock and is still active by the
// orchestrator's own accounting, exactly like `delivering`, `waiting_human`,
// `waiting_ai_review` and `repairing`, none of which are listed here either.
// The console's TERMINAL_STATUSES answers a narrower, display-only question
// -- "is this row done enough that a human should stop counting it toward
// the active/needs-attention totals" -- and `pipeline_failed` is a
// deterministic, bounded hand-off (status.go: "a brief hand-off to the
// repair-or-block decision") that resolves to `repairing` or `blocked`
// before an operator can meaningfully act on it as "in flight" work; leaving
// it out of this set is what let pipeline_failed runs show as permanently
// active in the sidebar (#360).
export const TERMINAL_STATUSES = new Set(["completed", "blocked", "failed", "cancelled", "pipeline_failed"]);
/**
 * Terminal runs whose thread is drawn cut rather than finished, and whose
 * event-feed tone reads as a failure rather than a success (see
 * overview.tsx's feedTone). `pipeline_failed` belongs here for the same
 * reason it belongs in TERMINAL_STATUSES above: it names a failed pipeline
 * command, and treating it as a plain "terminal" run without also treating
 * it as a "cut" one would have `feedTone`/`isTerminal` fall through to their
 * completed-run branch and paint a failed run's thread and feed entry green.
 */
export const CUT_STATUSES = new Set(["blocked", "failed", "cancelled", "pipeline_failed"]);
/** The set the "Needs you" triage and the nav's hot count are built from. */
export const NEEDS_ATTENTION_STATUSES = new Set(["waiting_human", "blocked"]);

export function statusMeta(status: string): StatusMeta {
  return STATUS_META[status] ?? { label: status.replaceAll("_", " "), variant: "idle", pulse: false, phase: null };
}

export const isTerminal = (status: string): boolean => TERMINAL_STATUSES.has(status);

// The meters' denominators. The API reports how much of each budget a run has
// spent but not the cap, so the console carries one: these are the caps the
// retired Python orchestrator enforced (RetryBudget), kept as the display
// budget because the Go V1 orchestrator has no retry policy — it neither caps
// these counters nor increments any of them (see `reachedPhase`).
export const ATTEMPT_BUDGETS = {
  planning: 2,
  implementation: 3,
  pipelineRepair: 3,
  reviewCycles: 3,
  ciRepair: 3,
  totalExecutions: 10,
} as const;

export type AttemptRow = { label: string; used: number; budget: number; emphasis?: boolean };

export function attemptRows(workflow: Workflow): AttemptRow[] {
  return [
    { label: "Planning", used: workflow.planningAttempts, budget: ATTEMPT_BUDGETS.planning },
    { label: "Implementation", used: workflow.implementationAttempts, budget: ATTEMPT_BUDGETS.implementation },
    { label: "Pipeline repair", used: workflow.pipelineRepairAttempts, budget: ATTEMPT_BUDGETS.pipelineRepair },
    { label: "Review cycles", used: workflow.reviewCycles, budget: ATTEMPT_BUDGETS.reviewCycles },
    { label: "CI repair", used: workflow.ciRepairAttempts, budget: ATTEMPT_BUDGETS.ciRepair },
    { label: "Total executions", used: workflow.totalAgentExecutions, budget: ATTEMPT_BUDGETS.totalExecutions, emphasis: true },
  ];
}

const phaseIndex = (phase: Phase): number => PHASES.indexOf(phase);

/**
 * How far along the phase path a run demonstrably got.
 *
 * `workflow_runs.current_phase` tracks the status column and is overwritten with
 * `blocked`/`failed`/`cancelled` when a run ends badly, so it cannot say where a
 * terminal run stopped. The attempt counters and the pull request can: each one
 * is only ever written by the phase that owns it. Everything here is therefore
 * evidence of a phase having been entered, never an assumption.
 *
 * The Go V1 orchestrator increments planningAttempts, reviewCycles,
 * ciRepairAttempts and pipelineRepairAttempts, but never implementationAttempts
 * (server.go has no writer for it), so a run that never reached review or a
 * repair cycle -- most of a happy-path `preparing`/`delivering` run's life --
 * has nothing in the counters above to place it past `plan`; the live status
 * below is what carries this for those statuses, exactly as it does for
 * `offered`.
 *
 * Specification task A2 replaces this with gate state derived server-side; this
 * function is the seam that change lands on.
 */
export function reachedPhase(workflow: Workflow): number {
  let reached = 0;
  if (workflow.planningAttempts > 0) reached = Math.max(reached, phaseIndex("plan"));
  if (workflow.implementationAttempts > 0) reached = Math.max(reached, phaseIndex("implement"));
  if (workflow.pipelineRepairAttempts > 0) reached = Math.max(reached, phaseIndex("pipeline"));
  if (workflow.reviewCycles > 0) reached = Math.max(reached, phaseIndex("review"));
  if (workflow.pullRequestUrl) reached = Math.max(reached, phaseIndex("pr"));
  if (workflow.ciRepairAttempts > 0 || workflow.pullRequestState) reached = Math.max(reached, phaseIndex("checks"));
  if (workflow.status === "completed") return phaseIndex("done");

  const live = statusMeta(workflow.status).phase;
  if (live) reached = Math.max(reached, phaseIndex(live));
  return reached;
}

export type GateState = "passed" | "failed" | "pending" | "not_reached";

export type Gate = { label: string; state: GateState; pendingLabel: string };

/** The five delivery gates, derived from `reachedPhase` (see its caveat). */
export function deriveGates(workflow: Workflow): Gate[] {
  const reached = reachedPhase(workflow);
  const cut = CUT_STATUSES.has(workflow.status);

  const gate = (label: string, phase: Phase, pendingLabel: string): Gate => {
    const index = phaseIndex(phase);
    if (reached > index) return { label, state: "passed", pendingLabel };
    if (reached < index) return { label, state: "not_reached", pendingLabel };
    // The run stopped exactly here: a cut run failed this gate, a live one is
    // still working on it.
    return { label, state: cut ? "failed" : "pending", pendingLabel };
  };

  return [
    gate("Plan valid", "plan", "planning"),
    gate("Local pipeline", "pipeline", "running"),
    gate("AI review", "review", "reviewing"),
    gate("GitHub checks", "checks", "pending"),
    workflow.status === "completed"
      ? { label: "Human approval", state: "passed", pendingLabel: "waiting" }
      : gate("Human approval", "human", "waiting"),
  ];
}

export const GATE_LABEL: Record<GateState, string> = {
  passed: "✓ passed",
  failed: "✗ failed",
  pending: "in progress",
  not_reached: "not reached",
};

// --- Events ---------------------------------------------------------------

/**
 * One human-readable line per workflow event.
 *
 * The events API returns the raw `event_type` and payload; specification task A3
 * moves this rendering to the orchestrator so every client agrees on the wording.
 * Until then it lives here.
 *
 * `persistExecutionEvent` (orchestrator/internal/server/server.go) stores the
 * runner's payload flat and unwrapped — there is no envelope to unwrap. The Go
 * vocabulary is `started`/`log`/`progress`/`completed`/`failed`/`cancelled`
 * (runner/internal/control/events.go's `validEventType`) plus
 * `workflow_transition`, `pull_request.created`, `pull_request.merged` and
 * `delivery.failed` (orchestrator/internal/server/{server,delivery}.go). The
 * Python writer's envelope and its recovery event types
 * (`offer_unanswered`/`lease_recovery_offered`/`execution_requeued`) were
 * retired with it in #247 and are not read here: no writer in the current
 * system produces them, so a fallback for that shape would be dead code.
 */
export function executionError(event: WorkflowEvent): string | null {
  if (event.type !== "failed") return null;
  return text(asRecord(event.payload).error) || null;
}

export function describeEvent(event: WorkflowEvent): { text: string; phase: string; warn: boolean } {
  const payload = asRecord(event.payload);

  switch (event.type) {
    case "workflow_transition": {
      // The Go writer's only producer today (controlWorkflow's retry branch)
      // writes `{"reason": ...}` with no `status` field, so the `status`-keyed
      // branch below is dead against real payloads; `reason` is what carries
      // the operator-facing text and is rendered directly when present.
      const status = text(payload.status);
      const reason = text(payload.reason);
      return {
        text: reason || (status ? `Workflow entered ${statusMeta(status).label.toLowerCase()}` : "Workflow state changed"),
        phase: status ? shortPhase(status) : "workflow",
        warn: status === "blocked" || status === "failed",
      };
    }
    case "started":
      // The runner's started payload is just `{"status":"running"}`
      // (control_loop.go's `execute`) — no runner or job identity travels with
      // it, and the events API doesn't surface either alongside the payload
      // (see `WorkflowEvent` in api.ts), so there is nothing more to report.
      return { text: "Agent execution started", phase: "execution", warn: false };
    case "progress":
      return { text: text(payload.message) || "Agent reported progress", phase: "execution", warn: false };
    case "log": {
      // The timeline reads as sentences, so the escape sequences the agent wrote
      // for a terminal are dropped here rather than rendered. The agent log pane
      // is where they are drawn in colour (ui/ansi.tsx).
      const line = logText(event.payload);
      return { text: line === null ? "Agent log output" : stripAnsi(line), phase: "log", warn: false };
    }
    case "completed":
      return { text: "Agent execution completed", phase: "execution", warn: false };
    case "failed": {
      // `terminalPayload` (runner/internal/dispatch/control_loop.go) always
      // sets `exitCode`, but guard `undefined` anyway rather than rendering
      // "exit undefined" if a future writer omits it.
      const exit = payload.exitCode;
      return {
        text: exit === undefined ? "Agent execution failed" : `Agent execution failed (exit ${String(exit)})`,
        phase: "execution",
        warn: true,
      };
    }
    case "cancelled":
      return { text: "Agent execution cancelled", phase: "execution", warn: true };
    case "pull_request.created": {
      // delivery.go's INSERT writes `{"number": "<pr.Number>", "url": "..."}`.
      const number = text(payload.number);
      return { text: number ? `Pull request #${number} opened` : "Pull request opened", phase: "pr", warn: false };
    }
    case "pull_request.merged":
      return { text: "Pull request merged", phase: "merge", warn: false };
    case "plan.recorded": {
      // Written by preparePlanningCompletion (orchestrator/internal/server/
      // server.go) as `{"plan": [...]}` once a planner execution completes;
      // the first entry is always its summary (planFromPayload), the rest its
      // remaining-work steps.
      const plan = Array.isArray(payload.plan) ? (payload.plan as unknown[]).filter((entry) => typeof entry === "string") : [];
      return {
        text: plan.length > 0 ? `Plan ready: ${plan[0] as string}` : "Plan ready",
        phase: "plan",
        warn: false,
      };
    }
    case "delivery.failed": {
      // Written by `terminateWorkflow` with `{"reason": "<cause>"}`.
      const reason = text(payload.reason);
      return { text: reason ? `Delivery failed: ${reason}` : "Delivery failed", phase: "event", warn: true };
    }
    case "human.rejected": {
      // Written by `rejectWorkflow` (orchestrator/internal/server/delivery.go)
      // with `{"reason": "changes requested[: <comment>]"}` when an admin
      // rejects a run at the human-approval gate -- already self-describing,
      // unlike delivery.failed's bare cause, so it is rendered verbatim.
      const reason = text(payload.reason);
      return { text: reason || "Changes requested", phase: "human", warn: true };
    }
    default:
      return { text: event.type.replaceAll("_", " "), phase: "event", warn: false };
  }
}

/** The status vocabulary is long; the events column shows a compact stem. */
function shortPhase(status: string): string {
  const phase = statusMeta(status).phase;
  return phase ?? status;
}

/**
 * The text a `log` event carried. `EmitLog` (runner/internal/control/events.go)
 * writes `{"message": "<chunk>", "chunkIndex": …, "chunkCount": …}` flat, which
 * the top-level `message` key below covers; `log`/`text`/`line` are kept as
 * fallbacks for any other flat shape a future writer might use.
 */
export function logText(payload: unknown): string | null {
  if (typeof payload === "string") return payload;
  if (!payload || typeof payload !== "object") return null;
  const record = payload as Record<string, unknown>;
  for (const key of ["message", "log", "text", "line"]) {
    if (typeof record[key] === "string") return record[key] as string;
  }
  return null;
}

function asRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" ? (value as Record<string, unknown>) : {};
}

function text(value: unknown): string {
  return typeof value === "string" ? value : "";
}

File: orchestrator/internal/server/status.go
package server

import "slices"

// Status is the vocabulary written to app.workflow_runs.status. Before this
// type existed the vocabulary was spelled out as raw SQL string literals
// scattered across server.go, delivery.go and recovery.go, with no compiler
// check that a new writer used a value any reader recognised -- see #265.
// These eight values are every one those files write today; 021_workflow_run_
// status_check.sql (extended by 022_workflow_run_delivering_status.sql to add
// 'delivering') enforces the same set as a database CHECK, so the two copies
// cannot drift silently again.
//
// 'running' is deliberately not here: it is a status app.jobs uses for an
// execution in flight, but no code path ever writes it to a workflow run.
type Status string

const (
	// StatusOffered is a run's initial status: the scheduler has picked an
	// issue and offered its job to a runner, and nothing has answered yet.
	StatusOffered Status = "offered"
	// StatusPreparing is set once a runner accepts its job offer, and holds
	// for the whole execution -- there is no separate "running" status for a
	// workflow run, only for the job underneath it.
	StatusPreparing Status = "preparing"
	// StatusPlanning is set once a runner accepts a planning job offer (see
	// projectConfig.RequirePlanning), and holds for the whole planner
	// execution -- the same relationship StatusPreparing has to the developer
	// execution that follows it. ScheduleOnce dispatches a planner-role
	// packet instead of the developer packet for an opted-in project; once
	// that execution reports "completed", persistExecutionEvent records the
	// plan and re-offers the same job (app.jobs.workflow_run_id stays UNIQUE
	// -- this is a second offer/accept/lease cycle for the one job the
	// workflow already has, not a second job) with the developer packet,
	// carrying the plan forward as its Plan context. A planner execution that
	// fails, is blocked, or is cancelled falls through to the ordinary
	// terminal handling below, exactly like a failed developer execution.
	StatusPlanning Status = "planning"
	// StatusWaitingGithubChecks is set once a pull request has been opened
	// and the run is waiting for GitHub's checks to report a result.
	StatusWaitingGithubChecks Status = "waiting_github_checks"
	// StatusDelivering marks a run whose agent execution succeeded and whose
	// pull request is being opened. persistExecutionEvent sets it (in place of
	// StatusCompleted -- see #267) for the runner's 'completed' event, and it
	// deliberately keeps the project lock while it holds this status: the run
	// still has to hand its work to GitHub, and a crash before that finishes
	// must not look like nothing is holding the project.
	//
	// Before this status existed 'completed' was written at this point too,
	// which meant "the agent succeeded" and "the pull request merged" were the
	// same value and could only be told apart by whether a project_locks row
	// still happened to exist -- see resumeStrandedDeliveries (recovery.go)
	// for what that ambiguity cost. A run leaves 'delivering' for
	// 'waiting_github_checks' once deliverWorkflow opens the pull request, or
	// sideways to 'blocked' if opening it fails (blockExternal).
	StatusDelivering Status = "delivering"
	// StatusWaitingHuman marks a run whose GitHub checks are already green and
	// whose project opted into the human-approval gate (projectConfig's
	// RequireHumanApproval): observeWorkflow stops here instead of merging
	// automatically, and the run holds its project lock exactly like
	// StatusWaitingGithubChecks -- it is still doing work, just waiting on a
	// person instead of on GitHub. SubmitHumanDecision is what moves a run away
	// from this status: "approved" sends it back to StatusWaitingGithubChecks
	// so the same tested merge path runs again (checks are already green, so
	// the next observer tick, or SubmitHumanDecision's own best-effort
	// immediate call, merges it straight away); "changes_requested" sends it to
	// StatusBlocked with the reviewer's comment as the reason, the same
	// terminal shape an agent's own declared block uses.
	StatusWaitingHuman Status = "waiting_human"
	// StatusWaitingAiReview marks a run whose developer execution reported
	// success and whose project opted into the independent-AI-review gate
	// (projectConfig's EnableAiReview): persistExecutionEvent sets it in place
	// of StatusDelivering for that project, and dispatchReviewerJob
	// (review.go) reopens the run's one job for a second, independent
	// execution -- a fresh reviewer session with no access to the developer's
	// own conversation, per AGENTS.md's "use fresh context for independent AI
	// review". The run holds its project lock across this status exactly like
	// StatusWaitingHuman: it is still doing work, just work a second agent
	// is doing instead of the first.
	//
	// handleReviewCompletion (review.go) is what moves a run away from this
	// status once the reviewer's own terminal event arrives: an approving
	// verdict hands off to deliverWorkflow, the same path a project with AI
	// review disabled always used; a rejecting verdict (or a reviewer
	// execution that crashed without one) ends the run at StatusBlocked with
	// the verdict recorded in app.ai_reviews -- the same terminal shape a
	// failed deterministic pipeline check would use, and the signal #354's
	// repair loop is meant to consume.
	StatusWaitingAiReview Status = "waiting_ai_review"
	// StatusRepairing marks a run whose independent AI review rejected the
	// developer's own attempt, or whose own deterministic pipeline failed a
	// required command (#352), and whose project opted into the repair loop
	// (projectConfig's EnableRepairLoop): repairOrBlock/pipelineFailedOrBlock
	// (repair.go) set it, via dispatchRepairJob/dispatchPipelineRepairJob, in
	// place of terminating the run at StatusBlocked -- the same
	// bounded escape hatch #354 adds so a rejection or a failed pipeline is not
	// always the end of the run, spent from ci_repair_attempts or
	// pipeline_repair_attempts respectively (two independent counters sharing
	// one bound and one opt-in, so a project need not configure "retry a
	// rejected review" and "retry a failed pipeline" separately). The run holds
	// its project lock across this status exactly like StatusWaitingAiReview or
	// StatusPipelineFailed: it is still doing work, a second (or third)
	// developer attempt informed by the reviewer's findings or the pipeline's
	// own failing command instead of a blind re-run.
	//
	// A repaired attempt's own "completed" event is not read any differently
	// than the original developer attempt's: persistExecutionEvent's switch on
	// job role and event type does not consult the run's current status, so the
	// repaired attempt flows back through the exact same
	// StatusDelivering/StatusWaitingAiReview/StatusPipelineFailed branch the
	// first attempt did, until its own attempts column exhausts the bound and
	// the matching *OrBlock falls through to StatusBlocked for good.
	StatusRepairing Status = "repairing"
	// StatusPipelineFailed marks a run whose developer (or repair) execution's
	// agent succeeded but whose project's own configured local pipeline --
	// PROJECT.md's "deterministic completion gate" -- failed a required
	// command (#352): persistExecutionEvent sets it in place of StatusDelivering
	// (or StatusWaitingAiReview) for that event, and pipelineFailedOrBlock
	// (repair.go) decides right after commit whether to repair the run
	// (bounded by pipeline_repair_attempts, the same EnableRepairLoop opt-in
	// #354's AI-review repair already uses) or end it at StatusBlocked. The run
	// holds its project lock across this status exactly like
	// StatusWaitingAiReview: the pipeline's own verdict already exists (unlike
	// AI review, no second execution has to run before there is one to act
	// on), so this status is a brief hand-off to the repair-or-block decision,
	// not a wait for more information.
	StatusPipelineFailed Status = "pipeline_failed"
	// StatusCompleted marks a run whose pull request GitHub has confirmed
	// merged (observeWorkflow) -- the true terminal "done" state. See
	// StatusDelivering for the status this run passed through on the way
	// here, and genuinelyTerminalStatuses for why StatusCompleted is still
	// not treated as terminal by the retry/cancel guard even though nothing
	// today writes a run away from it.
	StatusCompleted Status = "completed"
	// StatusFailed marks an agent execution that ended badly and did not
	// declare itself blocked, or a job whose lease lapsed without a runner
	// reporting an outcome.
	StatusFailed Status = "failed"
	// StatusBlocked marks a run an agent deliberately stopped and explained,
	// or one external delivery (opening/merging the pull request) could not
	// complete.
	StatusBlocked Status = "blocked"
	// StatusCancelled marks a run an operator stopped, or one whose job
	// offer nobody answered.
	StatusCancelled Status = "cancelled"
)

// String satisfies fmt.Stringer so a Status prints as its bare value in log
// lines and test failure messages, the same as the string literals it
// replaces did.
func (s Status) String() string { return string(s) }

// knownStatuses backs ParseStatus. It is built from the constants above
// rather than restated so the parser and the vocabulary can never disagree.
var knownStatuses = map[Status]bool{
	StatusOffered:             true,
	StatusPreparing:           true,
	StatusPlanning:            true,
	StatusWaitingGithubChecks: true,
	StatusWaitingHuman:        true,
	StatusWaitingAiReview:     true,
	StatusRepairing:           true,
	StatusPipelineFailed:      true,
	StatusDelivering:          true,
	StatusCompleted:           true,
	StatusFailed:              true,
	StatusBlocked:             true,
	StatusCancelled:           true,
}

// ParseStatus validates a value read back from app.workflow_runs.status (or
// received over the wire) against the known vocabulary. The database CHECK
// constraint (021_workflow_run_status_check.sql) already refuses anything
// else at write time, so a false here on a row read from this database means
// the constraint and this list have drifted, not that the row is unusual.
func ParseStatus(value string) (Status, bool) {
	status := Status(value)
	return status, knownStatuses[status]
}

// terminalStatuses are the workflow-run statuses a run never leaves under
// its own status column's ordinary meaning. Both the Go predicate below and
// the SQL the active-workflow gauge counts with (020_metrics_indexes.sql's
// workflow_runs_active_idx) are derived from this one list: they were
// independent copies, and a fifth terminal status added to one and not the
// other would have left the gauge counting finished work as active, with
// nothing to catch it.
//
// StatusDelivering is deliberately absent: a run holding it is still doing
// active work (opening a pull request), the same reason 'offered', 'preparing',
// 'planning', 'waiting_github_checks', 'waiting_human', 'waiting_ai_review',
// 'pipeline_failed' and 'repairing' are absent -- a run running its planner execution holds the project lock
// exactly as actively as one running its developer execution, a run waiting
// on a person to decide, or dispatching a bounded repair attempt, is exactly
// as active as one waiting on GitHub's checks, and must keep its project lock
// and count toward moirai_active_workflows the same
// way. See genuinelyTerminalStatuses for the narrower set terminateWorkflow's
// own guard uses, and StatusDelivering's doc comment above for why 'completed'
// and 'delivering' no longer share one meaning the way this list's single
// StatusCompleted entry once had to cover.
var terminalStatuses = []Status{StatusCompleted, StatusFailed, StatusBlocked, StatusCancelled}

func terminalStatus(state string) bool {
	status, ok := ParseStatus(state)
	return ok && slices.Contains(terminalStatuses, status)
}

// genuinelyTerminalStatuses are the three statuses no code path ever moves a
// run away from once reached -- deliberately not the same set as
// terminalStatuses, which also lists StatusCompleted.
//
// StatusCompleted is left out here on purpose, even though (since #267 split
// 'completed' from StatusDelivering) nothing today actually moves a run away
// from it once GitHub confirms the merge: RetryWorkflow's guard is "current
// status is genuinely terminal", and a completed run is done, not retryable,
// so it must read as *not* genuinely terminal here to keep failing that guard
// with "not retryable" rather than being treated as eligible for a fresh
// attempt. StatusDelivering is excluded from this list for the opposite,
// ordinary reason: a run holding it is still in flight, and blockExternal (via
// terminateWorkflow) sideways-moves it to 'blocked' when opening or merging
// the pull request fails -- see StatusDelivering's doc comment. See
// terminalStatuses for the (different, and equally intentional) set that does
// include StatusCompleted.
var genuinelyTerminalStatuses = []Status{StatusFailed, StatusBlocked, StatusCancelled}

func genuinelyTerminalStatus(state string) bool {
	status, ok := ParseStatus(state)
	return ok && slices.Contains(genuinelyTerminalStatuses, status)
}
